query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -154,6 +186,65 @@ private void extractColumns(List<String> columns, Expression expression) {
     }
   }
 
+  private Duration parseDuration(String timeSeriesPeriod) {
+    String[] splitPeriodString = timeSeriesPeriod.split(":");
+    long amount = Long.parseLong(splitPeriodString[0]);
+    ChronoUnit unit = TimeUnit.valueOf(splitPeriodString[1]).toChronoUnit();
+    return Duration.of(amount, unit);
+  }
+
+  private Optional<Duration> findTimeRangeDuration(Filter filter, String timeFilterColumn) {
+
+    // time filter will always be present with AND operator
+    if (filter.getOperator() != Operator.AND) {
+      return Optional.empty();
+    }
+
+    Optional<Long> timeRangeStart =
+        filter.getChildFilterList().stream()
+            .filter(
+                childFilter ->
+                    this.isMatchingFilter(
+                        childFilter, timeFilterColumn, List.of(Operator.GE, Operator.GT)))
+            .map(matchingFilter -> matchingFilter.getRhs().getLiteral().getValue().getLong())
+            .findFirst();
+
+    Optional<Long> timeRangeEnd =
+        filter.getChildFilterList().stream()
+            .filter(
+                childFilter ->
+                    this.isMatchingFilter(
+                        childFilter, timeFilterColumn, List.of(Operator.LT, Operator.LE)))
+            .map(matchingFilter -> matchingFilter.getRhs().getLiteral().getValue().getLong())
+            .findFirst();
+
+    if (timeRangeStart.isPresent() && timeRangeEnd.isPresent()) {
+      return Optional.of(
+          Duration.ofSeconds(
+              TimeUnit.SECONDS.convert(
+                  timeRangeEnd.get() - timeRangeStart.get(), TimeUnit.MILLISECONDS)));
+    }
+
+    return filter.getChildFilterList().stream()
+        .map(childFilter -> this.findTimeRangeDuration(childFilter, timeFilterColumn))
+        .flatMap(Optional::stream)
+        .findFirst();
+  }
+
+  private boolean isMatchingFilter(Filter filter, String column, Collection<Operator> operators) {
+    return column.equals(filter.getLhs().getColumnIdentifier().getColumnName())
+        && (operators.stream()
+            .anyMatch(operator -> Objects.equals(operator, filter.getOperator())));
+  }
+
+  public void computeTimeRangeDuration(QueryRequest queryRequest) {
```

Comment:
I'm assuming this gets exposed because we can't do the calculation on construction (timeFilterColumn has to be set first). Instead though, we could just do this work in the getter so a caller doesn't need to remember to compute before accessing.

Reply:
correct - see initial comment. We have the query request, but not the time range. Later, the time range gets set. We don't need to give the query request again, or to explicitly call calculate before getting - instead, compute it just in time. The caller invokes the getter with no args which returns an optional duration - empty if it can't calculate it (either because it's missing or time range col unknown).

Reply:
CC @kotharironak